### ISSUE TITLE
New Project Dialog Slow for Java Templates

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#402]
@rat-moonshine and I noticed that the File > New > Java * Project actions were very slow on macOS.   For me, it takes several seconds for the form to open and close in this case.  

I experimented with creating some other projects, and I only saw problems on the Java templates.

**Comment by piotrzarzycki21** [comment]
Slow opening dialog for creating Java project was caused by property which is being marked as Bindable. W are setup that property when we are changing **Parent directory**, for some reason setup some paths make setup of that property really slow. Optimization for Bindable on property is in most of the cases using custom event and dispatch it whenever it is really needed. Following commit https://github.com/prominic/Moonshine-IDE/commit/c663d17f0816851ac7a3fedbb847c9aa6bf1fb49 is doing exactly that which get rid of slow launch of new java project. 
Let me know if now it's working correctly.

**Comment by rat-moonshine** [comment]
I noticed this problem while working on #458 today. I have updated latest commits from 'master'. Upon test on Windows, I still found waiting took 4-10 seconds when keep opening the dialog against both the Java projects in same application session. 

Since the waiting still slower than other projects, is it possible that Java projects doing something extra before opening the dialog?

**Comment by piotrzarzycki21** [comment]
I don't think so it is doing something extra. Did you work on Installed version on launched from IDE ?

**Comment by rat-moonshine** [comment]
I have launched by IDE but do you think something can make difference between installer and IDE-run? I shall try to give some test to this later some time.

**Comment by piotrzarzycki21** [comment]
> I have launched by IDE but do you think something can make difference between installer and IDE-run? I shall try to give some test to this later some time.

In terms of performance and slow working of course it may be some difference. You are running debug version of SWF from IDE. Installed version is non debug - At least I hope so!

**Comment by rat-moonshine** [comment]
Hmm.. I have downloaded the latest build from Bamboo,
![image](https://user-images.githubusercontent.com/24370870/53423282-18658800-3a07-11e9-9ff9-33e63296a3bc.png)

With installed version test, the waiting seconds improved to 4-6 seconds. Once it reached to 9 also but mostly were remain by 6 seconds for me on Windows.

**Comment by piotrzarzycki21** [comment]
Problem exists on Mac - Did you check there ?

**Comment by rat-moonshine** [comment]
No, only on Windows for now.

**Comment by piotrzarzycki21** [comment]
> No, only on Windows for now.
> How many seconds does it usually taking in your tests?

I didn't do such tests :) It was previously visible slowness. :) Do you experience that on Mac?

**Comment by rat-moonshine** [comment]
How you tested then to determine if the fix you applied is suppose to get rid of slow launch?

**Comment by piotrzarzycki21** [comment]
By clicking on new Java Maven Project. Eventually when I discovered what was all about by changing paths using combo box.

**Comment by piotrzarzycki21** [comment]
I see in the original description that you were able to reproduce that issue. Make the same steps as it was previously.

**Comment by JoelProminic** [comment]
File > New > Java Maven Project is still slow when the new project form opens.  I also noticed slowness when entering a new value for the project name.

It felt like File > New > Java Gradle Project took even longer to open the new project form.  I wasn't able to reproduce the slowness for updating the project name, though.

**Comment by piotrzarzycki21** [comment]
> File > New > Java Maven Project is still slow when the new project form opens. I also noticed slowness when entering a new value for the project name.
> 
> It felt like File > New > Java Gradle Project took even longer to open the new project form. I wasn't able to reproduce the slowness for updating the project name, though.

That's not good. I definitely felt difference after my changes. On my sight I saw even busy indicator from Mac that something happen. Once I pushed changes it was gone and everything was faster. Can you record short video?

**Comment by piotrzarzycki21** [comment]
I think I know what is slowing down all of that. It looks like I was able to reproduce some slowness on windows. Whenever we are trying to access [projectFolder](https://github.com/prominic/Moonshine-IDE/blob/3d20962df0826b2a9d03129390aa68225b2a0760/ide/MoonshineSharedCore/src/actionScripts/valueObjects/ProjectVO.as#L100) property [FileWrapper](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/valueObjects/FileWrapper.as). Inside of it we are doing [updateChildren](https://github.com/prominic/Moonshine-IDE/blob/3d20962df0826b2a9d03129390aa68225b2a0760/ide/MoonshineSharedCore/src/actionScripts/valueObjects/FileWrapper.as#L204) which calling [getDirectoryListing](https://github.com/prominic/Moonshine-IDE/blob/3d20962df0826b2a9d03129390aa68225b2a0760/ide/MoonshineSharedCore/src/actionScripts/valueObjects/FileWrapper.as#L231) - That part can slow everything. I believe we can do this one async, so there would be no delay on object creation. I will try to experiment with that.

**Comment by rat-moonshine** [comment]
[FileWrapper](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/valueObjects/FileWrapper.as) fourth argument if provide _false_, then its `updateChildren(..)` is not suppose to get call immediately. 

If `updateChildren(..)` only causing this slowness in case of Java, I'm wondering why did we choose to provide the said argument value _true_ during project dialog open when we may not need the project's complete file-system. Might worth to look inside [CreateProject](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/as3project/CreateProject.as) once.

**Comment by piotrzarzycki21** [comment]
> [FileWrapper](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/valueObjects/FileWrapper.as) fourth argument if provide _false_, then its `updateChildren(..)` is not suppose to get call immediately.
> 
> If `updateChildren(..)` only causing this slowness in case of Java, I'm wondering why did we choose to provide the said argument value _true_ during project dialog open when we may not need the project's complete file-system. Might worth to look inside [CreateProject](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/as3project/CreateProject.as) once.

That argument was probably provided to postpone heavy operations. I believe we should take different approach. First is get rid of that argument and have async way of collecting those children. Next we should have smart system of busy indicating that something longer is happen. With code which I have pushed you now can see some small delay on showing folders in sidebar. We won't avoid more heavy operations on folders trees. In the future we may event more complicated operations on startup done in sidebar. I think what we should do:
1. Have busy indicator when we are opening new project tab
2. Have busy indicator for each loading folder and for all projects during first run. Example how it looks like in Intellij 12. 
![initializing](https://user-images.githubusercontent.com/24554795/53489636-f9f89e80-3a91-11e9-8bd0-d8ab6859d7cd.png)

![loading](https://user-images.githubusercontent.com/24554795/53489658-067cf700-3a92-11e9-8bae-3d70b5acf9ed.png)



I would start from 1 as part of that issue.

**Comment by rat-moonshine** [comment]
> First is get rid of that argument

I would not recommend this as the argument used in different places in different conditions - where parsing file-system not requires at all (asynchronous or synchronous). So let's not remove this.

**Comment by piotrzarzycki21** [comment]
> > First is get rid of that argument
> 
> I would not recommend this as the argument used in different places in different conditions - where parsing file-system not requires at all (asynchronous or synchronous). So let's not remove this.

But with code which I have pushed there is no logic which was handle by that flag, or at least that logic work in a different way. I feel like code which is taking list of folders/files running so often shouldn't be used in a sync way. Like I said having that in the async way should be connected with some busy indicator.

**Comment by JoelProminic** [comment]
I don't have the slowness issue for any of the other projects (Flex Desktop, Visual Editor, etc.).  What do the New Java Project forms do differently, and why do they need different behavior for project creation?

**Comment by piotrzarzycki21** [comment]
> I don't have the slowness issue for any of the other projects (Flex Desktop, Visual Editor, etc.). What do the New Java Project forms do differently, and why do they need different behavior for project creation?

I have analyzed problem farther and I'm sure that issue is with setting up `projectFolder ` property and reading content of provided path. However the difference is that in all of that projects Flex Desktop, Visual Editor etc. we are providing initially path to template of the project -> 
`elements\templates\projects\Flex Desktop Project (MacOS, Windows)\`, which doesn't contains to much files so reading list of the folders is quite fast. In case of Java project we are providing path **Parent Directory** initially instead path to template.
![path](https://user-images.githubusercontent.com/24554795/53635724-4c66c600-3c1e-11e9-90ae-fa082268d1a6.png)

In this directory can be much more so if your provide path to directory which contains hundreds files/folders - here we have our freeze.

**Comment by piotrzarzycki21** [comment]
Despite what is happen inside JavaProjectVO or AS3ProjectVO I think we should take general different approach when we are creating project. Creating itself project may be quite heavy no matter what we will do to optimize it. What come to my thought is that we could take as an approach that we are creating project once user click button Create. In the moment when user is filling all require input - we could have simple VO object which just hold values - it would be really fast not heavy at all:

1. User click on new project ex. Java Maven Project
2. We are creating lightweight VO ex. `PreProjectVO`
3. User filling all required fields
4. User is clicking on Create and we are using `PreProjectVO `information to create project

I'm not sure if it's possible but this is what come to my mind.

**Comment by rat-moonshine** [comment]
> In case of Java project we are providing path **Parent Directory** initially instead path to template.

Why we're doing so? What would happen if we provide the specific path to the template instead of its parent directory?

**Comment by piotrzarzycki21** [comment]
> Why we're doing so? What would happen if we provide the specific path to the template instead of its parent directory?

I don't know, but it seems to me much more natural providing path to the real project instead of template. Why it points to template and when It's changing ?

**Comment by piotrzarzycki21** [comment]
Ok I don't have any other idea to resolve that issue, so unless we decide in which direction to go or let's stay open till users felt wrong with that.

**Comment by rat-moonshine** [comment]
I have a bit uneasyness with this particular project type handling, we have created many project types including you which includes Royale, Visual Editor etc. No project (thankfully) so far didn't shown noticeable slowness during project dialog opening. I really would like to see what is happening by going in deep at some time. I'll plan the investigation some time later unless you come up something that fix the issue before.

**Comment by piotrzarzycki21** [comment]
> I have a bit uneasyness with this particular project type handling, we have created many project types including you which includes Royale, Visual Editor etc. No project (thankfully) so far didn't shown noticeable slowness during project dialog opening. I really would like to see what is happening by going in deep at some time. I'll plan the investigation some time later unless you come up something that fix the issue before.

You have proposition of fixes/improvements in the previous comments. It's the matter of choose to which direction we will go, unless you find something more in your investigation.

**Comment by rat-moonshine** [comment]
> You have proposition of fixes/improvements in the previous comments. It's the matter of choose to which direction we will go, unless you find something more in your investigation.

I still don't know how this template toggling happening in case of Java projects. We already have template toggling existed against options like `ActionScript Project` etc. So I wonder what is the problem in pointing template directories instead to their parent, if that would save some process?

**Comment by piotrzarzycki21** [comment]
> I still don't know how this template toggling happening in case of Java projects. We already have template toggling existed against options like `ActionScript Project` etc. So I wonder what is the problem in pointing template directories instead to their parent, if that would help in slowness?

Absolutely no problem. - At least I don't see it. To me it's just weird solution - creating project, pointing to template, make some magic and later change path to proper one. :)

**Comment by rat-moonshine** [comment]
Actually the core logic process were existed since the original source from Moonshine authors. If you see, they were required to look into the template folder to determine some stuff (if I remember now correctly) before starting up the project creation dialog with specific options available for opening project type.

**Comment by piotrzarzycki21** [comment]
> Actually the core logic process were existed since the original source from Moonshine authors. If you see, they were required to look into the template folder to determine some stuff (if I remember now correctly) before starting up the project creation dialog with specific options available for opening project type or something.. I even can't think clearly now ;/

Ahh that's the case, so Java project probably doesn't have such "needs", that's why main folder is being pointed out, instead of template.

**Comment by JoelProminic** [comment]
This may be unrelated, but when I tested a new Visual Editor project today, I noticed that there was no default value for the Template:
![image](https://user-images.githubusercontent.com/24532731/54396984-d6814680-4682-11e9-9190-e6a6192cffe3.png)

If I did not select a template, I got this error when creating a project:

>TestVisualEditor_0314_NoTemplate.mxml does not exist on the filesystem.
>Operation canceled.

Could this form have been affected by the Java Project form updates?

**Comment by piotrzarzycki21** [comment]
> This may be unrelated, but when I tested a new Visual Editor project today, I noticed that there was no default value for the Template:
> ![image](https://user-images.githubusercontent.com/24532731/54396984-d6814680-4682-11e9-9190-e6a6192cffe3.png)
> 
> If I did not select a template, I got this error when creating a project:
> 
> > TestVisualEditor_0314_NoTemplate.mxml does not exist on the filesystem.
> > Operation canceled.
> 
> Could this form have been affected by the Java Project form updates?
> 
> UPDATE: this did not reproduce after a Moonshine restart - Moonshine may have been in a bad state because of my earlier tests.

I don't think that I broke something in that area. I will observe this one as well.

**Comment by piotrzarzycki21** [comment]
I have switched to the template path and that's fixed an issue.

**Comment by JoelProminic** [comment]
I tested this for new Java Maven and Java Gradle projects, and I confirmed that the performance felt similar to the other project templates.
